File: receiver/src/phoneshare_receiver/services/rules.py
"""`packages/shared/src/rules.ts` Python portu — davranissal olarak birebir ayni.

Control plane (TS) ve agent (Python) ayni girdi icin ayni hedef klasor/dosya adini
uretmek zorundadir; `tests/test_rules_parity.py` bunu rules.test.ts case'leriyle dogrular.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Literal

from ..core.normalize import (
    DAY_MS,
    get_base_name,
    get_extension,
    glob_to_regexp,
    normalize_text,
    parse_date,
    parse_size,
    split_list,
)

MatchType = Literal["extension", "filename", "source", "size", "date", "tag"]
ConflictPolicy = Literal["overwrite", "rename", "version", "skip"]

CONFLICT_POLICIES: frozenset[str] = frozenset({"overwrite", "rename", "version", "skip"})


@dataclass(frozen=True, slots=True)
class Folder:
    id: str
    name: str
    path: str
    parent_id: str | None = None
    is_default: bool = False

    @classmethod
    def from_api(cls, data: dict) -> Folder:
        return cls(
            id=str(data["id"]),
            name=str(data.get("name") or ""),
            path=str(data.get("path") or ""),
            parent_id=data.get("parentId"),
            is_default=bool(data.get("isDefault")),
        )


@dataclass(frozen=True, slots=True)
class Rule:
    id: str
    name: str
    priority: int
    enabled: bool
    match_type: MatchType
    match_value: str
    target_folder_id: str
    rename: str | None = None
    conflict_policy: ConflictPolicy = "rename"

    @classmethod
    def from_api(cls, data: dict) -> Rule:
        policy = data.get("conflictPolicy") or "rename"
        return cls(
            id=str(data["id"]),
            name=str(data.get("name") or ""),
            priority=int(data.get("priority") or 0),
            enabled=bool(data.get("enabled", True)),
            match_type=data["matchType"],
            match_value=str(data.get("matchValue") or ""),
            target_folder_id=str(data["targetFolderId"]),
            rename=data.get("rename") or None,
            conflict_policy=policy if policy in CONFLICT_POLICIES else "rename",
        )


@dataclass(frozen=True, slots=True)
class FileMeta:
    file_name: str
    size_bytes: int
    source_app: str | None = None
    mime_type: str | None = None
    created_at: str | datetime | None = None
    tags: tuple[str, ...] = field(default_factory=tuple)


@dataclass(frozen=True, slots=True)
class ResolvedTarget:
    folder_id: str | None
    file_name: str
    conflict_policy: ConflictPolicy
    matched_rule_id: str | None


# ------------------------------------------------------------------ #
# Tekil eslesticiler                                                  #
# ------------------------------------------------------------------ #


def _match_extension(file: FileMeta, value: str) -> bool:
    ext = get_extension(file.file_name)
    if not ext:
        return False
    file_name = re.split(r"[\\/]", file.file_name)[-1] or file.file_name
    for raw in split_list(value):
        pattern = raw.strip()
        if not pattern:
            continue
        if "*" in pattern or "?" in pattern:
            if glob_to_regexp(pattern).search(file_name):
                return True
            continue
        if re.sub(r"^\.", "", pattern).lower() == ext:
            return True
    return False


def _match_filename(file: FileMeta, value: str) -> bool:
    haystack = normalize_text(file.file_name)
    for raw in split_list(value):
        needle = normalize_text(raw)
        if not needle:
            continue
        if "*" in needle or "?" in needle:
            if glob_to_regexp(needle).search(haystack):
                return True
            continue
        if needle in haystack:
            return True
    return False


def _match_source(file: FileMeta, value: str) -> bool:
    if not file.source_app:
        return False
    haystack = normalize_text(file.source_app)
    return any(
        needle and needle in haystack for needle in (normalize_text(r) for r in split_list(value))
    )


_RANGE_RE = re.compile(r"^([^-]+)-(.+)$")
_SIZE_OP_RE = re.compile(r"^(>=|<=|>|<|=)?\s*(.+)$", re.DOTALL)


def _match_size(file: FileMeta, value: str) -> bool:
    raw = value.strip()
    range_m = _RANGE_RE.match(raw)
    if range_m and not re.match(r"^[<>=]", raw):
        low = parse_size(range_m.group(1))
        high = parse_size(range_m.group(2))
        if low is None or high is None:
            return False
        return low <= file.size_bytes <= high

    m = _SIZE_OP_RE.match(raw)
    if not m:
        return False
    op = m.group(1) or "="
    size_bytes = parse_size(m.group(2))
    if size_bytes is None:
        return False
    if op == ">":
        return file.size_bytes > size_bytes
    if op == ">=":
        return file.size_bytes >= size_bytes
    if op == "<":
        return file.size_bytes < size_bytes
    if op == "<=":
        return file.size_bytes <= size_bytes
    return file.size_bytes == size_bytes


_DATE_OP_RE = re.compile(r"^(>=|<=|>|<)\s*(.+)$", re.DOTALL)


def _match_date(file: FileMeta, value: str) -> bool:
    at = parse_date(file.created_at)
    if at is None:
        return False
    raw = value.strip()

    parts = raw.split("..")
    if len(parts) == 2:
        start = parse_date(parts[0].strip())
        end = parse_date(parts[1].strip())
        if start is None or end is None:
            return False
        # Bitis gunu dahil olsun diye gun sonuna tasinir.
        return start <= at <= end + DAY_MS - 1

    m = _DATE_OP_RE.match(raw)
    if m:
        bound = parse_date(m.group(2).strip())
        if bound is None:
            return False
        op = m.group(1)
        if op == ">":
            return at > bound
        if op == ">=":
            return at >= bound
        if op == "<":
            return at < bound
        return at <= bound + DAY_MS - 1

    day = parse_date(raw)
    if day is None:
        return False
    return day <= at < day + DAY_MS


def _match_tag(file: FileMeta, value: str) -> bool:
    tags = [normalize_text(t) for t in (file.tags or ())]
    if not tags:
        return False
    return any(normalize_text(raw) in tags for raw in split_list(value))


# ------------------------------------------------------------------ #
# Genel API                                                           #
# ------------------------------------------------------------------ #

_MATCHERS = {
    "extension": _match_extension,
    "filename": _match_filename,
    "source": _match_source,
    "size": _match_size,
    "date": _match_date,
    "tag": _match_tag,
}


def match_rule(file: FileMeta, rule: Rule) -> bool:
    """Tek bir kuralin dosyaya uyup uymadigini soyler. Devre disi kural asla eslesmez."""
    if not rule.enabled:
        return False
    if not rule.match_value or not rule.match_value.strip():
        return False
    matcher = _MATCHERS.get(rule.match_type)
    if matcher is None:
        return False
    return matcher(file, rule.match_value)


_UNSAFE_SEGMENT = re.compile('[\\/:*?"<>|\x00-\x1f]')


def sanitize_segment(value: str) -> str:
    """Dosya sistemi icin guvenli olmayan karakterleri temizler."""
    out = _UNSAFE_SEGMENT.sub("_", value)
    out = re.sub(r"\s+", " ", out).strip()
    return re.sub(r"[. ]+$", "", out)


def apply_rename_template(
    template: str,
    file: FileMeta,
    seq: int = 1,
    seq_padding: int = 3,
) -> str:
    """Sablon: {date} {source} {orig} {ext} {seq}.

    Ornek: "{date}_{source}_{orig}.{ext}" -> "2024-05-03_whatsapp_fatura.pdf"
    """
    at = parse_date(file.created_at)
    if at is None:
        at = int(datetime.now(tz=UTC).timestamp() * 1000)
    d = datetime.fromtimestamp(at / 1000, tz=UTC)
    values = {
        "date": f"{d.year:04d}-{d.month:02d}-{d.day:02d}",
        "source": normalize_text(file.source_app) if file.source_app else "unknown",
        "orig": get_base_name(file.file_name),
        "ext": get_extension(file.file_name),
        "seq": str(seq).zfill(seq_padding),
    }

    out = re.sub(
        r"\{(date|source|orig|ext|seq)\}",
        lambda m: values.get(m.group(1).lower(), ""),
        template,
        flags=re.IGNORECASE,
    )

    ext = values["ext"]
    # Sablonda {ext} yoksa orijinal uzantiyi koru.
    if (
        ext
        and not re.search(r"\{ext\}", template, re.IGNORECASE)
        and not out.lower().endswith(f".{ext}")
    ):
        out = f"{out}.{ext}"
    # Uzanti bos kaldiysa olusan sondaki noktayi temizle.
    out = re.sub(r"\.+$", "", out)

    sanitized = sanitize_segment(out)
    return sanitized if sanitized else file.file_name


def sort_rules(rules: list[Rule]) -> list[Rule]:
    """Oncelik (kucuk once), esitlikte id ile deterministik siralama."""
    return sorted(rules, key=lambda r: (r.priority, r.id))


def resolve_target(
    file: FileMeta,
    rules: list[Rule],
    folders: list[Folder],
    default_folder_id: str | None = None,
    default_conflict_policy: ConflictPolicy = "rename",
    seq: int = 1,
    seq_padding: int = 3,
) -> ResolvedTarget:
    """Oncelik sirasina gore ILK eslesen kural kazanir; yoksa varsayilan klasore duser."""
    folder_ids = {f.id for f in folders}

    for rule in sort_rules(rules):
        if not match_rule(file, rule):
            continue
        # Hedef klasor silinmisse kural gecersiz sayilir, sonraki kurala bakilir.
        if folders and rule.target_folder_id not in folder_ids:
            continue
        return ResolvedTarget(
            folder_id=rule.target_folder_id,
            file_name=(
                apply_rename_template(rule.rename, file, seq=seq, seq_padding=seq_padding)
                if rule.rename
                else file.file_name
            ),
            conflict_policy=rule.conflict_policy,
            matched_rule_id=rule.id,
        )

    fallback = default_folder_id
    if fallback is None:
        fallback = next((f.id for f in folders if f.is_default), None)
    if fallback is None:
        fallback = next((f.id for f in folders if f.parent_id is None), None)

    return ResolvedTarget(
        folder_id=fallback,
        file_name=file.file_name,
        conflict_policy=default_conflict_policy,
        matched_rule_id=None,
    )

File: packages/shared-types/src/rules.ts
import type {
  ConflictPolicy,
  FileMeta,
  Folder,
  ResolvedTarget,
  Rule,
} from "./types";
import {
  getBaseName,
  getExtension,
  globToRegExp,
  normalizeText,
  parseDate,
  parseSize,
  splitList,
} from "./normalize";

/* ------------------------------------------------------------------ */
/* Tekil eslesticiler                                                  */
/* ------------------------------------------------------------------ */

/**
 * "*.pdf", "pdf", ".PDF", "*.doc?" veya "pdf, docx" formatlarini destekler.
 * Bosluklu liste icindeki her parca ayri ayri denenir (OR).
 */
function matchExtension(file: FileMeta, value: string): boolean {
  const ext = getExtension(file.fileName);
  if (!ext) return false;
  const fileName = file.fileName.split(/[\\/]/).pop() ?? file.fileName;
  return splitList(value).some((raw) => {
    const pattern = raw.trim();
    if (!pattern) return false;
    // Tam glob verilmisse dosya adinin tamamiyla karsilastir: "*.pdf", "rapor*.xlsx"
    if (pattern.includes("*") || pattern.includes("?")) {
      return globToRegExp(pattern).test(fileName);
    }
    // Sade uzanti verilmisse: "pdf" veya ".pdf"
    return pattern.replace(/^\./, "").toLowerCase() === ext;
  });
}

/** Dosya adi icinde gecen metin; buyuk/kucuk harf ve Turkce karakter duyarsiz. */
function matchFilename(file: FileMeta, value: string): boolean {
  const haystack = normalizeText(file.fileName);
  return splitList(value).some((raw) => {
    const needle = normalizeText(raw);
    if (!needle) return false;
    if (needle.includes("*") || needle.includes("?")) {
      return globToRegExp(needle).test(haystack);
    }
    return haystack.includes(needle);
  });
}

/** Kaynak uygulama adi (WhatsApp, Safari, Mail...). Kismi eslesme kabul edilir. */
function matchSource(file: FileMeta, value: string): boolean {
  if (!file.sourceApp) return false;
  const haystack = normalizeText(file.sourceApp);
  return splitList(value).some((raw) => {
    const needle = normalizeText(raw);
    return needle.length > 0 && haystack.includes(needle);
  });
}

/**
 * "> 10MB", ">=1gb", "< 500KB", "= 1024", "10MB-20MB" formatlari.
 */
function matchSize(file: FileMeta, value: string): boolean {
  const raw = value.trim();
  const range = /^([^-]+)-(.+)$/.exec(raw);
  if (range && !/^[<>=]/.test(raw)) {
    const min = parseSize(range[1] as string);
    const max = parseSize(range[2] as string);
    if (min === null || max === null) return false;
    return file.sizeBytes >= min && file.sizeBytes <= max;
  }
  const m = /^(>=|<=|>|<|=)?\s*(.+)$/.exec(raw);
  if (!m) return false;
  const op = m[1] ?? "=";
  const bytes = parseSize(m[2] as string);
  if (bytes === null) return false;
  switch (op) {
    case ">":
      return file.sizeBytes > bytes;
    case ">=":
      return file.sizeBytes >= bytes;
    case "<":
      return file.sizeBytes < bytes;
    case "<=":
      return file.sizeBytes <= bytes;
    default:
      return file.sizeBytes === bytes;
  }
}

/**
 * "2024-01-01..2024-12-31" araligi, ">= 2024-01-01", "<= 2024-12-31"
 * veya tek gun "2024-05-03" (o gunun tamami).
 */
function matchDate(file: FileMeta, value: string): boolean {
  const at = parseDate(file.createdAt ?? null);
  if (at === null) return false;
  const raw = value.trim();

  const range = raw.split("..");
  if (range.length === 2) {
    const from = parseDate((range[0] as string).trim());
    const to = parseDate((range[1] as string).trim());
    if (from === null || to === null) return false;
    // Bitis gunu dahil olsun diye gun sonuna tasinir.
    return at >= from && at <= to + DAY_MS - 1;
  }

  const m = /^(>=|<=|>|<)\s*(.+)$/.exec(raw);
  if (m) {
    const bound = parseDate((m[2] as string).trim());
    if (bound === null) return false;
    switch (m[1]) {
      case ">":
        return at > bound;
      case ">=":
        return at >= bound;
      case "<":
        return at < bound;
      default:
        return at <= bound + DAY_MS - 1;
    }
  }

  const day = parseDate(raw);
  if (day === null) return false;
  return at >= day && at < day + DAY_MS;
}

const DAY_MS = 24 * 60 * 60 * 1000;

/** Etiketlerden herhangi biri eslesirse true. */
function matchTag(file: FileMeta, value: string): boolean {
  const tags = (file.tags ?? []).map(normalizeText);
  if (tags.length === 0) return false;
  return splitList(value).some((raw) => tags.includes(normalizeText(raw)));
}

/* ------------------------------------------------------------------ */
/* Genel API                                                           */
/* ------------------------------------------------------------------ */

/** Tek bir kuralin dosyaya uyup uymadigini soyler. Devre disi kural asla eslesmez. */
export function matchRule(file: FileMeta, rule: Rule): boolean {
  if (!rule.enabled) return false;
  if (!rule.matchValue || rule.matchValue.trim() === "") return false;
  switch (rule.matchType) {
    case "extension":
      return matchExtension(file, rule.matchValue);
    case "filename":
      return matchFilename(file, rule.matchValue);
    case "source":
      return matchSource(file, rule.matchValue);
    case "size":
      return matchSize(file, rule.matchValue);
    case "date":
      return matchDate(file, rule.matchValue);
    case "tag":
      return matchTag(file, rule.matchValue);
    default:
      return false;
  }
}

function pad(n: number, len: number): string {
  return String(n).padStart(len, "0");
}

/** Dosya sistemi icin guvenli olmayan karakterleri temizler. */
function sanitizeSegment(value: string): string {
  return value
    .replace(/[\\/:*?"<>|\u0000-\u001f]/g, "_")
    .replace(/\s+/g, " ")
    .trim()
    .replace(/[. ]+$/, "");
}

export interface RenameContext {
  /** {seq} icin sira numarasi (varsayilan 1). */
  seq?: number;
  /** {seq} icin basamak sayisi (varsayilan 3 -> 001). */
  seqPadding?: number;
}

/**
 * Yeniden adlandirma sablonu: {date} {source} {orig} {ext} {seq}
 * Ornek: "{date}_{source}_{orig}.{ext}" -> "2024-05-03_whatsapp_fatura.pdf"
 */
export function applyRenameTemplate(
  template: string,
  file: FileMeta,
  ctx: RenameContext = {},
): string {
  const at = parseDate(file.createdAt ?? null) ?? Date.now();
  const d = new Date(at);
  const date = `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1, 2)}-${pad(d.getUTCDate(), 2)}`;
  const ext = getExtension(file.fileName);
  const orig = getBaseName(file.fileName);
  const source = file.sourceApp ? normalizeText(file.sourceApp) : "unknown";
  const seq = pad(ctx.seq ?? 1, ctx.seqPadding ?? 3);

  const values: Record<string, string> = { date, source, orig, ext, seq };

  let out = template.replace(/\{(date|source|orig|ext|seq)\}/gi, (_m, key: string) => {
    return values[key.toLowerCase()] ?? "";
  });

  // Sablonda {ext} yoksa orijinal uzantiyi koru.
  if (ext && !/\{ext\}/i.test(template) && !out.toLowerCase().endsWith(`.${ext}`)) {
    out = `${out}.${ext}`;
  }
  // Uzanti bos kaldiysa olusan sondaki noktayi temizle.
  out = out.replace(/\.+$/, "");

  const sanitized = sanitizeSegment(out);
  return sanitized.length > 0 ? sanitized : file.fileName;
}

export interface ResolveOptions {
  /** Hicbir kural eslesmezse kullanilacak klasor id'si (folders icindeki isDefault'u ezer). */
  defaultFolderId?: string | null;
  /** Hicbir kural eslesmediginde uygulanacak politika. */
  defaultConflictPolicy?: ConflictPolicy;
  rename?: RenameContext;
}

/** Kurallari oncelik (kucuk once), sonra id'ye gore deterministik siralar. */
export function sortRules(rules: readonly Rule[]): Rule[] {
  return [...rules].sort((a, b) => {
    if (a.priority !== b.priority) return a.priority - b.priority;
    return a.id.localeCompare(b.id);
  });
}

/**
 * Oncelik sirasina gore ILK eslesen kural kazanir.
 * Hicbiri eslesmezse varsayilan klasore (isDefault ya da options.defaultFolderId) duser.
 */
export function resolveTarget(
  file: FileMeta,
  rules: readonly Rule[],
  folders: readonly Folder[],
  options: ResolveOptions = {},
): ResolvedTarget {
  const folderIds = new Set(folders.map((f) => f.id));

  for (const rule of sortRules(rules)) {
    if (!matchRule(file, rule)) continue;
    // Hedef klasor silinmisse kural gecersiz sayilir, sonraki kurala bakilir.
    if (folders.length > 0 && !folderIds.has(rule.targetFolderId)) continue;
    return {
      folderId: rule.targetFolderId,
      fileName: rule.rename
        ? applyRenameTemplate(rule.rename, file, options.rename)
        : file.fileName,
      conflictPolicy: rule.conflictPolicy,
      matchedRuleId: rule.id,
    };
  }

  const fallback =
    options.defaultFolderId ??
    folders.find((f) => f.isDefault)?.id ??
    folders.find((f) => f.parentId === null)?.id ??
    null;

  return {
    folderId: fallback,
    fileName: file.fileName,
    conflictPolicy: options.defaultConflictPolicy ?? "rename",
    matchedRuleId: null,
  };
}
